addons/pos_epson_printer/static/src/js/printers.js

```diff
@@ -2,98 +2,167 @@
+                        // Pixel on the bottom left
+                        errors[x - 1][y + 1] += 3/16 * error;
+                    }
+                    if (y < height - 1) {
+                        // Pixel below
+                        errors[x][y + 1] += 5/16 * error;
+                    }
+                    if (x < width - 1 && y < height - 1) {
+                        // Pixel on the bottom right
+                        errors[x + 1][y + 1] += 1/16 * error;
+                    }
+                }
+            }
         }
+
+        return rasterData.join('');
     },
 
-    callback_createDevice: function (deviceObj, errorCode) {
-        if (deviceObj === null) {
-            Gui.showPopup('ErrorPopup', {
-                'title': _t('Connection to the printer failed'),
-                'body':  _t('Please check if the printer is still connected. Error code: ') + errorCode,
-            });
-            return;
+    /**
+     * Base 64 encode a raster image
+     */
+    _encodeRaster: function (rasterData) {
+        var encodedData = '';
+        for(var i = 0; i < rasterData.length; i+=8){
+            var sub = rasterData.substr(i, 8);
+            encodedData += String.fromCharCode(parseInt(sub, 2));
         }
-        this.printer = deviceObj;
-        this.printer.onreceive = function(response){
-            if (!response.success) {
-                Gui.showPopup('ErrorPopup', {
-                    'title': _t('Epson ePOS Error'),
-                    'body':  _t('An error happened while sending data to the printer. Error code: ') + response.code,
-                });
-            }
-        };
+        return btoa(encodedData);
     },
 
     /**
-     * Create the print request for webPRNT from a canvas
+     * Create the raster data from a canvas
      * 
      * @override
      */
     process_canvas: function (canvas) {
-        if (this.printer) {
-            this.printer.addTextAlign(this.printer.ALIGN_CENTER);
-            this.printer.addImage(canvas.getContext('2d'), 0, 0, canvas.width, canvas.height);
-            this.printer.addCut();
-        }
+        var rasterData = this._canvasToRaster(canvas);
+        var encodedData = this._encodeRaster(rasterData);
+        return QWeb.render('ePOSPrintImage', {
+            image: encodedData,
+            width: canvas.width,
+            height: canvas.height,
+        });
     },
 
     /**
      * @override
      */
     open_cashbox: function () {
-        if (this.printer) {
-            this.printer.addPulse();
-            this.printer.send();
-        }
+        var pulse = QWeb.render('ePOSDrawer');
+        this.send_printing_job(pulse);
     },
 
     /**
      * @override
      */
-    send_printing_job: function () {
-        if (this.printer) {
-            this.printer.send();
-            return {
-                result: true
-            };
-        }
+    send_printing_job: function (img) {
+        return $.ajax({
```

Comment:
this looks fine in this context but cc @xmo-odoo for double check

Reply:
Looks fine to me, the address seems to be generated from the printer IP set in the POS config, and this is whatever custom protocol the printer uses so makes perfect sense to not use odoo's rpc.

Only remarks would be styllistic e.g. since this is being completely rewritten use an `async` function using shorthands, that sort of things, but that's more of an @aab-odoo thing.

Also not sure what's going to happen if the request is rejected e.g. the printer was shutdown or crashed just before the job was sent?

Reply:
You could indeed write
```js
send_printing_job: async function (img) {
    const res = await $.ajax({...});
    return $(res).find('response').attr('success') === 'true';
},
```

Reply:
or 
```js
async send_printing_job(img) {
    ...
```
no?

Reply:
yep, I didn't suggest that by consistency with the other functions, but this is indeed how we write new code

Reply:
As I modify 80% of the file, I've used this new notation and changed the other functions too